lib/ocpp/v201/smart_charging.cpp

```diff
@@ -402,4 +403,436 @@ void SmartChargingHandler::conform_validity_periods(ChargingProfile& profile) co
+
+                bool within_window =
+                    continue_time_arrow(temp_time, period_start_time, period_end_time, lowest_next_time);
+
+                EVLOG_debug << "get_next_temp_time> Profile #" << profile.id << " " << temp_time << " < "
+                            << period_end_time << " && " << period_end_time << " < " << lowest_next_time << " = "
+                            << within_window;
+
+                // Step 7 - Continue if not within final time window
+                if (within_window) {
+                    lowest_next_time = period_end_time;
+                    EVLOG_debug << "get_next_temp_time> Profile #" << profile.id << " " << lowest_next_time
+                                << " is new lowest_next_time";
+                } else {
+                    EVLOG_debug << "get_next_temp_time> Profile #" << profile.id << " " << lowest_next_time
+                                << " is current lowest_next_time NO CHANGE";
+                }
+
+                period_start_time = period_end_time;
+            }
+        }
+    }
+    return lowest_next_time;
+}
+
+std::optional<ocpp::DateTime>
+SmartChargingHandler::get_absolute_profile_start_time(const std::optional<ocpp::DateTime> startSchedule) {
+    std::optional<ocpp::DateTime> period_start_time;
+
+    if (startSchedule) {
+        period_start_time.emplace(ocpp::DateTime(floor<seconds>(startSchedule.value().to_time_point())));
+    } else {
+        EVLOG_warning << "Absolute profile with no startSchedule, this should not be possible";
+    }
+
+    return period_start_time;
+}
+
+std::optional<ocpp::DateTime>
+SmartChargingHandler::get_recurring_profile_start_time(const ocpp::DateTime& time,
+                                                       const std::optional<ocpp::DateTime> startSchedule,
+                                                       const std::optional<RecurrencyKindEnum> recurrencyKind) {
+    std::optional<ocpp::DateTime> period_start_time;
+
+    if (startSchedule) {
+        // TODO ?: What if startSchedule is not set? This use case is not handled in the 1.6 code
+        const ocpp::DateTime start_schedule =
+            ocpp::DateTime(std::chrono::floor<seconds>(startSchedule.value().to_time_point()));
+        int seconds_to_go_back;
+
+        if (recurrencyKind.value() == RecurrencyKindEnum::Daily) {
+            seconds_to_go_back =
+                SmartChargingHandler::determine_duration(start_schedule, time) % (HOURS_PER_DAY * SECONDS_PER_HOUR);
+        } else {
+            seconds_to_go_back =
+                SmartChargingHandler::determine_duration(start_schedule, time) % (SECONDS_PER_DAY * DAYS_PER_WEEK);
+        }
+        auto time_minus_seconds_to_go_back = time.to_time_point() - seconds(seconds_to_go_back);
+        period_start_time.emplace(ocpp::DateTime(time_minus_seconds_to_go_back));
+    } else {
+        EVLOG_warning << "Recurring profile with no startSchedule, this should not be possible";
+    }
```

Comment:
So, the only case where we wouldn't have a `startSchedule` is if something has gone seriously wrong. As mentioned above, if we know the profile is valid we don't need to double down on checking for/handling invalid cases.

Reply:
I prefer to avoid assuming that other parts of systems always work as expected. I would prefer to reject this recommendation, and discuss refactorings at a later date.